src/main/java/org/allenai/ml/sequences/crf/CRFTrainer.java

```diff
@@ -0,0 +1,119 @@
+        val stateSpace = StateSpace.buildFromSequences(justLabels, startState, stopState);
+        logger.info("StateSpace: num states {}, num transitions {}",
+            stateSpace.states().size(), stateSpace.transitions().size());
+        double featAcceptProb = opts.minExpectedFeatureCount >= 1 ? 1.0/opts.minExpectedFeatureCount : 1.0;
+        val featOpts = CRFFeatureEncoder.BuildOpts.builder()
+            .numThreads(opts.numThreads)
+            .probabilityToAccept(featAcceptProb)
+            .build();
+        List<List<O>> unlabeledData = labeledData.stream()
+            .map(labeledDatum -> labeledDatum.stream().map(Pair::getOne).collect(Collectors.toList()))
+            .collect(Collectors.toList());
+        this.featureEncoder = CRFFeatureEncoder.build(unlabeledData, predicateExtractor, stateSpace, featOpts);
+        logger.info("Number of node predicates: {}, edge predicates: {}",
+            featureEncoder.nodeFeatures.size(), featureEncoder.edgeFeatures.size());
+        this.weightEncoder = new CRFWeightsEncoder<>(stateSpace,
+            featureEncoder.nodeFeatures.size(),
+            featureEncoder.edgeFeatures.size());
+    }
+
+    private void ensureStartStopPadded(List<List<S>> justLabels, S startState, S stopState) {
+        if (!justLabels.stream().allMatch(lst ->
+            lst.get(0).equals(startState) &&
+                lst.get(lst.size()-1).equals(stopState))) {
+            throw new IllegalArgumentException("Not all states padded with start/stop");
+        }
+    }
+
+    public CRFModel<S, O, F> modelForWeights(Vector weights) {
+        return new CRFModel<>(featureEncoder, weightEncoder, weights);
+    }
+
+    /**
+     * Train a CRFModel from labeled data and a predicateExtractor
+     * @param labeledData Assume that each sequence is padded with start/stop states
+     * @return Trained CRFModel
+     */
+    public CRFModel<S, O, F> train(List<List<Pair<O, S>>> labeledData)
+    {
+        ensureStartStopPadded(labeledData.stream()
+                .map(x -> x.stream().map(Pair::getTwo).collect(Collectors.toList()))
+                .collect(Collectors.toList()),
+            featureEncoder.stateSpace.startState(),
+            featureEncoder.stateSpace.stopState());
+        ExampleObjectiveFn<CRFIndexedExample> objective = new CRFLogLikelihoodObjective<>(weightEncoder);
+        List<CRFIndexedExample> indexedData = labeledData.stream()
+            .map(featureEncoder::indexLabeledExample)
+            .collect(toList());
+        val mrOpts = Parallel.MROpts.withThreads(opts.numThreads);
+        BatchObjectiveFn<CRFIndexedExample> objFn =
+            new BatchObjectiveFn<>(indexedData, objective, weightEncoder.numParameters(), mrOpts);
+        GradientFn regularizer = Regularizer.l2(objFn.dimension(), opts.sigmaSq);
+        val cachedObjFn = new CachingGradientFn(opts.lbfgsHistorySize, objFn.add(regularizer));
+        val quasiNewton = QuasiNewton.lbfgs(opts.lbfgsHistorySize);
```

Comment:
0 means gradient descent. 






On Mon, Jul 27, 2015 at 11:07 AM -0700, "Christopher Clark" <notifications@github.com> wrote:












In src/main/java/org/allenai/ml/sequences/crf/CRFTrainer.java:
> +    {
> +        ensureStartStopPadded(labeledData.stream()
> +                .map(x -> x.stream().map(Pair::getTwo).collect(Collectors.toList()))
> +                .collect(Collectors.toList()),
> +            featureEncoder.stateSpace.startState(),
> +            featureEncoder.stateSpace.stopState());
> +        ExampleObjectiveFn<CRFIndexedExample> objective = new CRFLogLikelihoodObjective<>(weightEncoder);
> +        List<CRFIndexedExample> indexedData = labeledData.stream()
> +            .map(featureEncoder::indexLabeledExample)
> +            .collect(toList());
> +        val mrOpts = Parallel.MROpts.withThreads(opts.numThreads);
> +        BatchObjectiveFn<CRFIndexedExample> objFn =
> +            new BatchObjectiveFn<>(indexedData, objective, weightEncoder.numParameters(), mrOpts);
> +        GradientFn regularizer = Regularizer.l2(objFn.dimension(), opts.sigmaSq);
> +        val cachedObjFn = new CachingGradientFn(opts.lbfgsHistorySize, objFn.add(regularizer));
> +        val quasiNewton = QuasiNewton.lbfgs(opts.lbfgsHistorySize);



Does this imply train only supports LBFGS not vanilla gradient descent? Or would a history of 0 for LBFGS be equal to vanilla gradient descent?



—
Reply to this email directly or view it on GitHub.